### ISSUE TITLE
Remove pyqt5-related dependencies

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-tasks-core#286]
EDIT: renamed this issue to (finally?) fix it for good, since #288 seems promising (after a few updates to napari-workflow-related packages)

---

When trying to collect the 0.7.0 tasks with server 1.0.2, I run into the qt installation issue on Mac again. This means I can, unfortunately, not test the real packages locally anymore.

I get this error:
```
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [29 lines of output]
      Traceback (most recent call last):
        File "/Users/joel/Dropbox/Joel/FMI/Code/fractal_test_setup/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.7.0/venv/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 144, in prepare_metadata_for_build_wheel
          hook = backend.prepare_metadata_for_build_wheel
      AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel'
      
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "/Users/joel/Dropbox/Joel/FMI/Code/fractal_test_setup/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.7.0/venv/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 351, in <module>
          main()
        File "/Users/joel/Dropbox/Joel/FMI/Code/fractal_test_setup/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.7.0/venv/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 333, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/Users/joel/Dropbox/Joel/FMI/Code/fractal_test_setup/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal-tasks-core0.7.0/venv/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 148, in prepare_metadata_for_build_wheel
          whl_basename = backend.build_wheel(metadata_directory, config_settings)
        File "/private/var/folders/sk/fl32m0g97kv2xlqz9f0p5bbm0000gr/T/pip-build-env-f6blv1g2/overlay/lib/python3.8/site-packages/sipbuild/api.py", line 46, in build_wheel
          project = AbstractProject.bootstrap('wheel',
        File "/private/var/folders/sk/fl32m0g97kv2xlqz9f0p5bbm0000gr/T/pip-build-env-f6blv1g2/overlay/lib/python3.8/site-packages/sipbuild/abstract_project.py", line 87, in bootstrap
          project.setup(pyproject, tool, tool_description)
        File "/private/var/folders/sk/fl32m0g97kv2xlqz9f0p5bbm0000gr/T/pip-build-env-f6blv1g2/overlay/lib/python3.8/site-packages/sipbuild/project.py", line 585, in setup
          self.apply_user_defaults(tool)
        File "project.py", line 69, in apply_user_defaults
          super().apply_user_defaults(tool)
        File "/private/var/folders/sk/fl32m0g97kv2xlqz9f0p5bbm0000gr/T/pip-build-env-f6blv1g2/overlay/lib/python3.8/site-packages/pyqtbuild/project.py", line 70, in apply_user_defaults
          super().apply_user_defaults(tool)
        File "/private/var/folders/sk/fl32m0g97kv2xlqz9f0p5bbm0000gr/T/pip-build-env-f6blv1g2/overlay/lib/python3.8/site-packages/sipbuild/project.py", line 236, in apply_user_defaults
          self.builder.apply_user_defaults(tool)
        File "/private/var/folders/sk/fl32m0g97kv2xlqz9f0p5bbm0000gr/T/pip-build-env-f6blv1g2/overlay/lib/python3.8/site-packages/pyqtbuild/builder.py", line 69, in apply_user_defaults
          raise PyProjectOptionException('qmake',
      sipbuild.pyproject.PyProjectOptionException
      [end of output]
```

Pointing to the pyqtbuild & qmake. Thus, reinforces again that it's important if we can avoid the napari & qt dependency for our core tasks.

**Comment by jluethi** [comment]
I can reproduce this error when just doing a `pip install pyqt5` in a fresh environment. But using conda and the conda-forge channel works:
`conda install pyqt`
Installs pyqt correctly

**Comment by jluethi** [comment]
Quick update: On my computer at home, this same installation runs through without an issue. I don't really understand why. Both are Mac's (the one at home on macOS Ventura, the work laptop on macOS Monterey).

Anyhow, let's still get rid of the qt dependency as soon as we can to avoid this. But I have a test setup now to run some more tests locally ;)

**Comment by tcompa** [comment]
> When trying to collect the 0.7.0 tasks with server 1.0.2, I run into the qt installation issue on Mac again. This means I can, unfortunately, not test the real packages locally anymore.
> 
> I get this error:

Which command originally triggered this error? Was it through `fractal task collect`?

(this is just for the record, since you already said that `pip install pyqt5` is enough to reproduce the error)

**Comment by jluethi** [comment]
Yes, it's triggered by `fractal task collect fractal-tasks-core --package-version 0.7.0`

**Comment by tcompa** [comment]
> Anyhow, let's still get rid of the qt dependency as soon as we can to avoid this.

Status update
* https://github.com/haesleinhuepf/napari-tools-menu/pull/10 was merged and released in [napari-tools-menu=0.1.18](https://github.com/haesleinhuepf/napari-tools-menu/releases/tag/0.1.18).
* https://github.com/haesleinhuepf/napari-skimage-regionprops/pull/38 is still open, and in principle it should allow importing `napari-skimage-regionprops` without napari/QT (probably some details of that PR still need to be fixed though).
* A similar PR would be needed for `napari-segment-blobs-and-things-with-membranes`.
* Do other dependencies depend on QT or napari? To be checked.

**Comment by jluethi** [comment]
Great to see progress on this.

As briefly discussed at the call today, even if those libraries can work without importing napari, they currently still have napari as a dependency. As long as it's core napari only, i.e. without `pyqt5`, this may be less of a problem for the mac installation though. `pip install napari` actually runs through successfully on both my macs (even the one that can't collect the fractal tasks locally), because it doesn't pick a qt version to install. 
So as long as we can drop the pyqt5 dependency, that would help. (napari doesn't actually work without some qt backend being installed. But that's fine by us, we don't need to start napari anymore, especially if the napari imports are handled that way)

I actually wonder why we have a specific pyqt5 dependency at the moment. Was it because napari imports would otherwise fail? Or do we have a specific pyqt5 dependency somewhere?

**Comment by tcompa** [comment]
Current status:
* It should be OK to import `napari-tools-menu` without QT bindings available;
* The same should be true for `napari-skimage-regionprops`, as soon as https://github.com/haesleinhuepf/napari-skimage-regionprops/pull/51 is merged&released;
* Something similar should be done for https://github.com/haesleinhuepf/napari-segment-blobs-and-things-with-membranes/blob/main/napari_segment_blobs_and_things_with_membranes/__init__.py.

Ongoing tests are in PR #288

**Comment by tcompa** [comment]
As of napari-skimage-regionprops=0.8.1, we now can safely import it without QT bindings.

Side comment, another consequence of updating napari-tools-menu to 0.1.19 is that now also`import napari_segment_blobs_and_things_with_membranes` only raises a warning for missing QT bindings, rather than an error.

That's why we now got all tests passing for #288 (just with a lot of warnings). If we add more napari-workflows packages, we'll have to make sure that they can be safely imported.

---

As a byproduct, removing pyqt5 packages means slightly smaller environments ( ~200 MB less), although the size remains quite large (especially due to pytorch):
```
$ du -hs lib/python3.9/site-packages/* | sort -h | tail -n 20
25M	lib/python3.9/site-packages/numba
26M	lib/python3.9/site-packages/numcodecs
28M	lib/python3.9/site-packages/OpenGL
32M	lib/python3.9/site-packages/babel
34M	lib/python3.9/site-packages/numpy
35M	lib/python3.9/site-packages/numpy.libs
35M	lib/python3.9/site-packages/scipy.libs
36M	lib/python3.9/site-packages/skimage
39M	lib/python3.9/site-packages/imagecodecs.libs
39M	lib/python3.9/site-packages/matplotlib
41M	lib/python3.9/site-packages/nbclassic
61M	lib/python3.9/site-packages/pandas
63M	lib/python3.9/site-packages/opencv_python_headless.libs
71M	lib/python3.9/site-packages/cv2
71M	lib/python3.9/site-packages/imageio_ffmpeg
73M	lib/python3.9/site-packages/imagecodecs
86M	lib/python3.9/site-packages/scipy
99M	lib/python3.9/site-packages/llvmlite
201M	lib/python3.9/site-packages/PyQt5
1.6G	lib/python3.9/site-packages/torch
```

**Comment by jluethi** [comment]
Oh, that's awesome. Once this is merged, let's release a 0.7.1 task package with it and I'll check whether local installation on M1 Macs then works :)

**Comment by tcompa** [comment]
There is now `0.7.1a0` on pypi, ready for testing.

**Comment by jluethi** [comment]
I tested local installation of `0.7.1a0` and now I'm not running into any issues with qt anymore. Unfortunately, there are still issues with `imagecodecs`.

I still need to test whether the examples run fine on `0.7.1a0`. Will do so on UZH servers once I have access there again.

I remember earlier last year, I had to remove pyqt5 & imagecodecs to libraries from the dependencies to get local installation of the tasks working, so this dependency saga should continue too much further. See discussion here: https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/289
